### ISSUE TITLE
Cheevos support

### ISSUE THREAD

**Issue by leiradel** [barbudreadmon/fbalpha2018#182]
Hi @barbudreadmon 

I took a look in the source code and this core doesn't support achievements. Is it possible to implement the `RETRO_ENVIRONMENT_SET_MEMORY_MAPS` environment call, or the `retro_get_memory_data/size` functions?

**Comment by barbudreadmon** [comment]
After looking at other cores, it seems to me like it's somehow relying on the same data than savestates. It could be a serious issue considering this data can change a lot from a release to another. Someone can confirm ?

**Comment by fr500** [comment]
It's not really the savestate data, but the SRAM data.

`retro_get_memory_data/size` aren't a good fit because those force SRAM saving to work via retroarch. Also for either case it would be a pita to map everything properly to all the supported boards.

For instance I know to handle saving in CPS2 we hacked the eeprom file so it saves that file in RETRO_GET_SAVE_DIRECTORY

With those functions we would instead get a pointer to the memory that should be saved and RetroArch would handle saving.

**Comment by barbudreadmon** [comment]
I think "BurnAreaScan(ACB_MEMORY_RAM)" handle saving only the sram data, perhaps @dinkc64 could confirm ?

@fr500 i removed the hack in the eeprom file a few months ago, it wasn't needed anymore.

**Comment by leiradel** [comment]
WRT achievements, most of them rely on RAM alone, but some may need SRAM when available. However, they never depend on savestates, just on being able to access RAM and SRAM.

Also, **all** the RAM must be accessible, i.e. some systems have some "quick" or "page 0" RAM that is easier/faster to access, and more RAM available on a different address. Both have to be mapped to enable achievements.

On systems where there's just one contiguous block of RAM, implementing `retro_get_memory_data/size` for RAM and SRAM is enough to support achievements. On systems with more complex memory layouts, `RETRO_ENVIRONMENT_SET_MEMORY_MAPS` must be used to map all the necessary block of memory.

Also, it's mandatory that the pointers returned and/or registered are stable between `retro_load_game` and `retro_unload_game`, since they are cached by the achievements code to speed up things.

**Comment by leiradel** [comment]
Forgot to say that, event for simple memory layouts where `retro_get_memory_data/size` can be used, it's preferable to use `RETRO_ENVIRONMENT_SET_MEMORY_MAPS` because they give more information about the memory than just it's pointer in the host memory and its size.

**Comment by fr500** [comment]
@barbudreadmon well thing is, if we use `retro_get_memory_data` that enables the frontend's mandatory handling of sram, which could be good I guess, as long as `BurnAreaScan(ACB_MEMORY_RAM` has all the data we need. Then we can remove the I/O based save handling.

The `RETRO_ENVIRONMENT_SET_MEMORY_MAPS` API is better for cheevos imho.

**Comment by leiradel** [comment]
We can have both.

**Comment by fr500** [comment]
Yes but we don't want to have two methods of saving persistent game data (we already save without retro_get_memory_data)

**Comment by leiradel** [comment]
I don't know anything about that, my request is to add achievements support

**Comment by barbudreadmon** [comment]
Answer from dink : 
>ACB_MEMORY_RAM is just for the ram,
ACB_NVRAM is for nvram
If you're lucky, the main ram will be the first block at the beginning of
ACB_MEMORY_RAM. some drivers might save the ram in the driver data (ACB_VOLATILE).  due to the fact that none of this was ever standardized..

**Comment by leiradel** [comment]
> If you're lucky, the main ram will be the first block at the beginning of ACB_MEMORY_RAM

Does it mean there could be more than one block of RAM, and that there's no way to tell if that's the case?

**Comment by barbudreadmon** [comment]
Well, an arcade machine has several block of ram, check the cps2 specs : https://en.wikipedia.org/wiki/CP_System_II
In the best case scenario, the *scan functions of the cpus and devices related to a driver will handle `BurnAreaScan(ACB_MEMORY_RAM)` properly and it will save all the rams and only the rams.

Actually the same can be said about console : some of them (if not all ?) had main ram + video ram + sound ram. Cheevos only needs main ram or it also needs video/sound ram ?

**Comment by fr500** [comment]
> I don't know anything about that, my request is to add achievements support
@leiradel I understand, but what I'm saying here is that the implementation should use the MEMORY_MAPS API, not the other one because it would introduce saving issues otherwise.

**Comment by leiradel** [comment]
> Cheevos only needs main ram or it also needs video/sound ram ?

Only main and save RAM, although all memory visible to the main CPU could be mapped via `RETRO_ENVIRONMENT_SET_MEMORY_MAPS`, including ROM.

Memory visible only for other CPUs can't be made visible because libretro doesn't have an API for that.

**Comment by leiradel** [comment]
> I understand, but what I'm saying here is that the implementation should use the MEMORY_MAPS API, not the other one because it would introduce saving issues otherwise.

Ah I see. Sure, I'm not saying what API should be used. If `retro_memory_get_data/size` will introduce issues, `RETRO_ENVIRONMENT_SET_MEMORY_MAPS` is the way to go.

**Comment by barbudreadmon** [comment]
Ok, let's take an example : the neogeo driver.

At line 278 of `src/burn/drv/neogeo/neo_run.cpp` i think there are all the things you need : 
- Neo68KRAM is main processor ram => **Needed ?**
- NeoZ80RAM is co-processor ram (should be sound) => **Unneeded ?**
- NeoNVRAM is the non volatile ram so i guess it is the "save ram" you mentioned => **Needed ?**
- NeoNVRAM2 is the second non volatile ram which is used only by gambling games => **Needed ?**
- NeoMemoryCard is the memory card => **Unneeded ?**

NB : lines 275 and 283 are the 3 video rams => **Unneeded ?**

You can get Neo68KRAM by getting the first 65536 bytes of `BurnAreaScan(ACB_MEMORY_RAM)`, and you can get NeoNVRAM (+ NeoNVRAM2 if present) with `BurnAreaScan(ACB_NVRAM)`, it should be 65536 bytes (+ 8192 bytes for the second one if present)

Also, the proper way to detect if you are running a neogeo game should be this boolean : `(BurnDrvGetHardwareCode() & (HARDWARE_PUBLIC_MASK - HARDWARE_PREFIX_CARTRIDGE)) == HARDWARE_SNK_NEOGEO`

I hope this is enough information if someone want to try and implement this on neogeo as a first step, feel free to ask if you need more information about the codebase.

**Comment by leiradel** [comment]
> Neo68KRAM is main processor ram => Needed ?

Needed.

> NeoZ80RAM is co-processor ram (should be sound) => Unneeded ?

Not needed, though I wonder if they could enable some achievements that would be otherwise impossible to make without it. Anyway, there's no way to report memory for CPUs other than the main CPU, unless this memory is mapped to the address space of the main CPU. In this case, they could be reported via the `RETRO_ENVIRONMENT_SET_MEMORY_MAPS` interface.

> NeoNVRAM is the non volatile ram so i guess it is the "save ram" you mentioned => Needed ?

It depends. If it just holds a real time clock or settings, no, it's not needed for achievements, but could be reported via `RETRO_MEMORY_RTC` so its state is maintained in disk.

> NeoNVRAM2 is the second non volatile ram which is used only by gambling games => Needed ?

Are there dumps of gambling games ROMs? If yes, maybe it's needed, depending on what's in it, but we can only report one NVM via `RETRO_MEMORY_RTC`. Everything can be reported using `RETRO_ENVIRONMENT_SET_MEMORY_MAPS` though. However, only the one reported via `RETRO_MEMORY_RTC` will be preserved by RetroArch.

> NeoMemoryCard is the memory card => Unneeded ?

Is this where savegames are stored? If so, then it's needed.

> NB : lines 275 and 283 are the 3 video rams => Unneeded ?

Same thing as the Z80 memory, considering this memory belongs to a VDP.

> you can get NeoNVRAM (+ NeoNVRAM2 if present)

Ah, so they contiguous in memory when NVRAM2 is present? Good.

> it should be 65536 bytes (+ 8192 bytes for the second one if present)

How can I be sure? RetroArch needs a pointer to the beginning of the memory, and its size.

Thanks for the help.

**Comment by barbudreadmon** [comment]
Yes i think nvram is for dipswitch settings, and probably hiscores and rtc.
There are a few gambling game on neogeo (V-Liner, Jockey GrandPrix)
The role of the memory card is kinda unclear to me, you can read about it there : http://www.neogeoprotos.com/memcardfaq.htm 
You can check if the game is a gambling game with `((BurnDrvGetHardwareCode() & HARDWARE_SNK_CONTROLMASK) == HARDWARE_SNK_GAMBLING)`

**Comment by barbudreadmon** [comment]
Also, BurnAreaScan won't do since you need a pointer.
Actually, if you can write some callback functions which can be called with the right args (pointers, sizes) from the driver, and handle reporting to RETRO_ENVIRONMENT_SET_MEMORY_MAPS, i think it would help a lot.

**Comment by barbudreadmon** [comment]
Not sure i made myself clear so i pushed some code, let me know if you think we can work out something from this

**Comment by salvadorc17** [comment]
BurnAreaScan() Dont seems to work for other system like Cps. Is not able to return the memory used, any way to get that one?

**Comment by leiradel** [comment]
@salvadorc17 this seems to be a different problem, I believe you should open a different issue.

**Comment by fr500** [comment]
AFAICT the callback was only added for NeoGEO for testing.
Once a solution is in place it could be implemented in other drivers.

**Comment by leiradel** [comment]
Hey guys, achievements are working with FBAlpha, thank you very much!

Just one more request: can the `RETRO_ACHIEVEMENTS` macro be removed and achievements support be always on?

Also, we can have achievements for any FBA driver now, it's just a matter of implementing the callback :+1:

**Comment by barbudreadmon** [comment]
I enabled it. Also, i'll probably do some refactoring on the way the callback is called, dink gave me some good ideas to make things in a less hectic way.

**Comment by salvadorc17** [comment]
Im interested in wich method can be used for calling retroachievements callback without touching drivers, im doing test in standalone fba.

**Comment by leiradel** [comment]
I believe the callback and macro should be changed to better reflect what this really is. Despite the motivation being to enable achievements, what this does is to report the System RAM to the libretro frontend, thats all. I.e. this could also be used for cheats.

**Comment by barbudreadmon** [comment]
Someone can confirm neogeo and cps1 RA are still working after my last commit ?

**Comment by salvadorc17** [comment]
Seems working fine for those, but now cps2 games does crash, will need extra code

**Comment by barbudreadmon** [comment]
Probably because i didn't reimplement cps2, you told me they weren't working properly in your last reply

**Comment by salvadorc17** [comment]
My mistake they were working, just seemed some memory areas was not properly aligned.

**Comment by meleu** [comment]
@barbudreadmon @leiradel 

I've just tested Metal Slug X.

The cheevos feature works on commit 8ad726a96c322a2742aa3e61d46121ebb54290ef
![mslugx-cheevo-58546](https://user-images.githubusercontent.com/8508804/37569349-c5c398f0-2abf-11e8-978b-30f20cf8e55a.png)

It doesn't work on commit 5a19dd2a33a0986aad5ee2050bdced823115240e

Do you guys need the `git bisect` thing or you already know where the problem was inserted?

**Comment by barbudreadmon** [comment]
What about the other neogeo games ? @salvadorc17 said they were working.

**Comment by meleu** [comment]
@barbudreadmon I'm not sure what he meant by "working". The games are working, but the cheevos feature doesn't. And Metal Slug X is the only arcade game with an achievement set on our site, [as you can see here](http://retroachievements.org/gameList.php?c=27).

**Comment by meleu** [comment]
Also, there's another thing that can lead to confusion: The RetroArch by itself is the piece of code responsible to load the achievements list. But the core needs to support cheevos correctly in order to make the achievements trigger.

When people load a game and open the Achievment List, they can believe that the cheevos feature is working fine, but they need to actually get a cheevo to confirm. That's what I did on my tests.

**Comment by barbudreadmon** [comment]
Do you mind telling me what changed in the data you are receiving ? Is the new data unusable or something ? Looks to me like what i'm sending after the refactoring from https://github.com/libretro/fbalpha/commit/3978cdeae2ab7cc8dd62c960102edd1da5ed6a64 should be the same as before, but perhaps i'm wrong.

**Comment by meleu** [comment]
are you on discord? if yes, look for meleu (me) at libretro server. I think we can talk better there

**Comment by barbudreadmon** [comment]
I'm not, and my oral english is lacking, so please explain here what you think the difference is compared to before https://github.com/libretro/fbalpha/commit/3978cdeae2ab7cc8dd62c960102edd1da5ed6a64 .

**Comment by salvadorc17** [comment]
Dont want to be contradictive but meleu is right, seems memory have changed his alignement for neogeo and new changes are not making cps2 work properly, so to be clear the previus first changes was working as best.

**Comment by barbudreadmon** [comment]
Can we find a workaround for the new implementation ?
If we stay with the new code, i don't mind having support for the 500+ drivers in the long term since all the code will be in the same place, if we go back to the old one where i have to hack every driver file, support will be limited to only a select few drivers.
I leave the choice up to you.

**Comment by fr500** [comment]
I agree with @barbudreadmon

**Comment by meleu** [comment]
@barbudreadmon 

> If we stay with the new code, i don't mind having support for the 500+ drivers in the long term since all the code will be in the same place

Oh yeah! This is a pretty strong argument! And I agree. As this is a very new feature I think we can rework what we did at RetroAchievements and from now on use the new implementation. 

@leiradel do you have something to comment about it?

---

@fr500 

The hash used for arcade on RALibretro is just the ROM name with no extension, then `mslugx` is:
```
$ echo -n mslugx | md5sum
320a36c48c06dce51438108df9defb90  -
```
I think we can do it on RetroArch's `cheevos.c`.

**Comment by leiradel** [comment]
This is not something we can adapt to in Retro Achievements or RetroArch, this is something that just works or not. I didn't have the chance to test it yet.

**Comment by meleu** [comment]
I thought it was just a matter of memory addresses being shifted. I'll ask the cheevos developers community to hold on with NeoGeo/Arcade games until we can solve/test this.

**Comment by fr500** [comment]
@leiradel I understand that, getting a log of the reported memory addresses and memory sizes before and after the change would be helpful to debug this.

**Comment by meleu** [comment]
[On my tests yesterday](https://github.com/libretro/fbalpha/issues/182#issuecomment-374027702), I was testing RetroArch and trying to get cheevos.

Today I tested on RALibretro itself. And when using the fbalpha 43ad83084606e6735adc9dd5db5b157028f92861 RALibretro crashes right after opening the Memory Inspector.

How to reproduce:

1. get [RALibretro pack here](http://bit.ly/RALibRetro).
2. get the latest fbalpha 32bit here: https://buildbot.libretro.com/nightly/windows/x86/latest/
3. unzip the `fbalpha_libretro.dll.zip` and put the file in the RALibretro's folder, inside the `Cores` folder.
4. launch RALibretro and load fbalpha
5. load a Neo Geo game
6. after checking the game is running fine, go to `RetroAchievements` -> `Memory Inspector`
7. **CRASH**

**Comment by salvadorc17** [comment]
How do you commit each separated build?? Will be needed to get back to the old method, new one was moving the memory in 0x4000 filled with zeros at start for NeoGeo.

**Comment by barbudreadmon** [comment]
> Will be needed to get back to the old method, new one was moving the memory in 0x4000 filled with zeros at start for NeoGeo.

Data from the first implementation is offsetted by 0x4000 or the 0x4000 first bytes are missing and replaced by 0 ? Or something else ?
If it's the first, that means there could be some kind of junk inserted at the beginning of the buffer when using the new method, so we could probably get back the data from the old implementation by using an additional offset on the new one.

**Comment by leiradel** [comment]
@meleu can you attach the RALibretro logs here, one with the latest core and one with a core that was working before?

**Comment by meleu** [comment]
@leiradel 

RALibretro's log when using a "good" fbalpha (bb651e3)
[good-log.txt](https://github.com/libretro/fbalpha/files/1829414/good-log.txt)

RALibretro's log when using the latest fbalpha (43ad830) **and opening the Memory Inspector (CRASH!)**
[crashlog.txt](https://github.com/libretro/fbalpha/files/1829412/crashlog.txt)

RALibretro's log when using the latest fbalpha (43ad830) **and not opening the Memory Inspector (no crash)**
[no-meminspector-log.txt](https://github.com/libretro/fbalpha/files/1829417/no-meminspector-log.txt)

**Comment by salvadorc17** [comment]
So the question is whats wrong with old method, it was working fine, and did make it usable for the most important arcade system.

**Comment by meleu** [comment]
@salvadorc17 [answer here](https://github.com/libretro/fbalpha/issues/182#issuecomment-374116603).

**Comment by barbudreadmon** [comment]
In no-meminspector-log.txt i see : 

> Moved recent file 0 to front "mslugx" - 13 - 27
Registered       0 bytes at 0x00000000
Installed        0 bytes at 0x00000000

Does that mean the core is reporting a size of 0 with a buffer filled with 0 ? If so then either my conditions are wrong (which would be weird, that kind of condition is all over the code), or retro_get_memory_data/retro_get_memory_size are called before the game is initialized (which is done in retro_load_game)
CPS1/CPS2 are also reporting a size of 0 with a buffer filled with 0 ?

**Comment by fr500** [comment]
It's called after load game but before the first retro_run

**Comment by meleu** [comment]
@barbudreadmon 

> CPS1/CPS2 are also reporting a size of 0 with a buffer filled with 0 ?

I think yes. Check the log:
[nocrash-cps1-log.txt](https://github.com/libretro/fbalpha/files/1829866/nocrash-cps1-log.txt)

**Comment by leiradel** [comment]
It's also called after `retro_load_game` but before `retro_run` in RALibretro.

**Comment by fr500** [comment]
Yes the API indicates it must be called before the first retro_run call

**Comment by barbudreadmon** [comment]
@fr500 even if the buffer is empty at frame 0, retro_get_memory_size should report a size of `0x00010000` if my `nHardwareFlag == HARDWARE_SNK_NEOGEO` condition is true.

**Comment by leiradel** [comment]
This is a problem for CPS1 and CPS2:

```
src/burner/libretro/libretro.cpp: In function 'void* retro_get_memory_data(unsigned int)':
src/burner/libretro/libretro.cpp:1847:24: warning: pointer of type 'void *' used in arithmetic [-Wpointer-arith]
    return (state_ptr + offset);
                        ^~~~~~
```

I'm testing Neo Geo now.

**Comment by leiradel** [comment]
`nHardwareFlag` has a bogus value of -2063532032 so it fails all the `if`s:

```
Moved recent file 0 to front "mslug" - 13 - 27
retro_get_memory_size -2063532032
retro_get_memory_data -2063532032
retro_get_memory_size -2063532032
Registered       0 bytes at 0x00000000
Installed        0 bytes at 0x00000000
retro_get_memory_size -2063532032
```

The value in hexadecimal is `0x85010000`. `0x85000000` doesn't correspond to any prefix in `burn.h` so `BurnDrvGetHardwareCode` is returning an invalid value in `retro_get_memory_data/size`.

**Comment by barbudreadmon** [comment]
Yeah, i noticed and was working on it, seems like some system actually needs something like `(nHardwareCode & (HARDWARE_PUBLIC_MASK - HARDWARE_PREFIX_CARTRIDGE))` instead of `(nHardwareCode & HARDWARE_PUBLIC_MASK)`, neogeo was one of them.

My last commit should fix both issues, i hope RA will work now.

**Comment by leiradel** [comment]
Also, The Punisher is returning `HARDWARE_CAPCOM_CPS1_QSOUND`, so maybe we need a mask of `0x7F000000`?

**Comment by salvadorc17** [comment]
Now i found cause of issue, memory data is added but never updated. So this method will not work because is not updating the reported ram. With the previous method it worked because the run functions was allowing that for each driver.

**Comment by barbudreadmon** [comment]
Ok, what about the new one ?

**Comment by meleu** [comment]
@barbudreadmon your last commit makes RALibretro "freeze" right after loading a game.

Here are the logs:

**CPS1**: [cps1-freeze-log.txt](https://github.com/libretro/fbalpha/files/1832785/cps1-freeze-log.txt)

**NeoGeo**: [neogeo-freeze-log.txt](https://github.com/libretro/fbalpha/files/1832786/neogeo-freeze-log.txt)

**Comment by leiradel** [comment]
I'll only be able to test later today, but `mainram_ptr` is initialized to `NULL` at the top of `burn_get_mainram_ptr_cb`, and then `memcpy(mainram_ptr, pba->Data, pba->nLen)` is executed just below. This will cause a crash, or am I missing something?

Also, even if `mainram_ptr` points to a block of memory at least the size of `pba->nLen`, won't we get a *copy* of the System RAM, frozen in time at the instant `memcpy` is executed?

**Comment by leiradel** [comment]
He, I've just seen @meleu 's comment. Yeah, I believe it's the issue with `mainram_ptr`.

**Comment by barbudreadmon** [comment]
What about now ?

**Comment by meleu** [comment]
@barbudreadmon @leiradel 

https://www.youtube.com/watch?v=MdN0NXgjsn8

**Comment by barbudreadmon** [comment]
@meleu Even the previous achievements from mslugx work ?

**Comment by meleu** [comment]
didn't check trying to earn them. I only checked the addresses and they seem to be fine. will make more tests later.

**Comment by meleu** [comment]
@barbudreadmon good news: yeah, it's working as expected!! Huge thanks for that!

Look the latest fbalpha commit hash on the bottom left "signature":
![fbalpha-works](https://user-images.githubusercontent.com/8508804/37713630-40b7fbb8-2cf6-11e8-9361-5eec431db451.png)

**Comment by leiradel** [comment]
@meleu everything is ok in RALibretro too, right?

**Comment by meleu** [comment]
@leiradel yep yep. RALibretro is running fine with fbalpha 8eabf37

**Comment by barbudreadmon** [comment]
@meleu @leiradel Nice, i'll make some changes to how size is handled this evening (i'll retrieve it from `pba->nLen` in burn_get_mainram_ptr_cb, it will make the code in retro_get_memory_size simpler).

**Comment by leiradel** [comment]
Awsome @barbudreadmon let us know when the changes are in so we can test again.

Should we keep this issue open as new drivers get support for achievements, or should we open one issue for each driver the community asks for achievements support?

**Comment by barbudreadmon** [comment]
Let me know if my changes broke something. I guess we can keep this issue open for future request.

**Comment by barbudreadmon** [comment]
Also removed the cast to UINT8 now that we don't need arithmetic anymore.

**Comment by meleu** [comment]
I'm going to make some tests now.

**Comment by barbudreadmon** [comment]
I did one last change, it doesn't seem to me like `BurnAreaScan` was needed in `retro_get_memory_data` as long as it is called once in `fba_init`

**Comment by meleu** [comment]
@barbudreadmon I've launched a bunch of games, for Neo Geo, CPS1, CPS2 and CPS3.

Neo Geo, CPS-1 and CPS-2 are running 100% fine.

CPS-3 is fine but when I open the Memory Inspector RALibretro crashes. Here are the crash logs:

playing "Street Fighter III: 3rd Impact" and then open Memory Inspector: [cps3-sfiii3-freeze-meminspector-log.txt](https://github.com/libretro/fbalpha/files/1838848/cps3-sfiii3-freeze-meminspector-log.txt)

playing "JoJo's Bizarre Adventure" and then open Memory Inspector: [cps3-jojoba-crash-meminspector-log.txt](https://github.com/libretro/fbalpha/files/1838847/cps3-jojoba-crash-meminspector-log.txt)

**Comment by barbudreadmon** [comment]
CPS-3 is not implemented yet.

**Comment by leiradel** [comment]
I'll add CPS1 and 2 to the list of supported drivers

**Comment by meleu** [comment]
Alright then.

A curiosity: the memory size is the same for NeGeo, CPS1 and CPS2?

All of them show the same amount of RAM when opening the Memory Inspector (it says "aware of 65536 RAM locations").

**Comment by barbudreadmon** [comment]
> A curiosity: the memory size is the same for NeGeo, CPS1 and CPS2?

Yes it is, their hardware is extremely similar.

CPS-3 will be 512k, and that's probably one of the biggest in fba, i'll add the support this week-end

**Comment by meleu** [comment]
@barbudreadmon is this serious?! e502e80

**Comment by barbudreadmon** [comment]
@meleu Why ? Is there something wrong with cps3 ?

**Comment by leiradel** [comment]
@barbudreadmon I think that @meleu is just excited about having CPS3 support, and not that he's dismissing the implementation. It's hard to communicate using a foreign language.

**Comment by meleu** [comment]
oh, sorry about the confusion. @leiradel is right, I'm excited! :sweat_smile: 

Is that commit adding support for CPS-3?

**Comment by barbudreadmon** [comment]
@meleu it should, let me know if there is any issue

**Comment by meleu** [comment]
@barbudreadmon oh yeah!! it's working fine!!! Opened Memory Inspector and everything seems to be working fine!

**Comment by Ntemis** [comment]
@barbudreadmon i know this is irrelevant somehow but i felled the urge to ask about 
Capcom ZN1/ZN2, Namco System 11/12 and Taito FX1, support possible?

**Comment by fr500** [comment]
FBA doesn't emulate any of those, please keep the discussion on-topic, this
is not a forum

**Comment by barbudreadmon** [comment]
Today i heard there was some incompatibility with RetroAchievements in neogeo games when changing bioses. Is that true ? Any idea what could be the issue ? Let me know if i can help with it.

**Comment by meleu** [comment]
@barbudreadmon it's intentional. We enforce the achievement developers to not allow their achievements trigger when using UniBIOS. The reason is on our docs:
http://docs.retroachievements.org/Console-Specific-Tips/#mandatory-unibios-protection

I'm not sure if it should be worked out on your end... Is it possible for the core know if the Achievements feature is enabled on RetroArch? Does it worth to spend your time coding such cheat protection in the core?

Maybe it's better let we at RetroAchievements deal with it and on your end just let users know about the protection we use.

**Comment by leiradel** [comment]
Libretro cores know nothing about achievements so it must be worked around on Retro Achievements.

Maybe we could deal with this in RALibretro but cheaters could just use RetroArch.

**Comment by barbudreadmon** [comment]
Ok, i don't see any problem if it's intentional then.

**Comment by fr500** [comment]
Maybe on detection disable hardcore mode :)

**Comment by meleu** [comment]
If going through this way I would suggest disable the Achievements feature as a whole. Even in softcore things like infinite lives/grenades/credits/whatever shouldn't be allowed.

**Comment by barbudreadmon** [comment]
I guess there is no way for the core to detect if retroachievements are enabled ? It would allow me to disable unibios.

**Comment by barbudreadmon** [comment]
@meleu Yes i did, it was requested for the new cheat search feature in retroarch. There are probably still a lot of drivers not exposing their ram though.

**Comment by meleu** [comment]
Hi @barbudreadmon .
This repo was renamed to fbalpha2018 and its descriptions says "**Deprecated** port of Final Burn Alpha to Libretro"

Now I'm worried if the current [libretro/fbalpha](https://github.com/libretro/fbalpha) repo/Core supports cheevos. If not, I need to make the RetroAchievements community aware that they must use fbalpha2018 for achievements.

**Comment by barbudreadmon** [comment]
> Now I'm worried if the current libretro/fbalpha repo/Core supports cheevos.

It does

**Comment by barbudreadmon** [comment]
Closing this issue, if there are other drivers you need with cheevos support, let me know in the new repo.
